tests/test_convert_text_to_mds.py

```diff
@@ -0,0 +1,185 @@
+                                  maybe_create_object_store_from_uri: Mock,
+                                  tmp_path: pathlib.Path, processes: int):
+    remote_folder = os.path.join(tmp_path, 'remote')
+    text_content = 'HELLO WORLD'
+    tokenizer_name = 'mosaicml/mpt-7b'
+    n_text_files = processes * 3
+
+    mock_object_store = Mock(
+        wraps=MockObjectStore(remote_folder, n_text_files, text_content))
+    maybe_create_object_store_from_uri.return_value = mock_object_store
+    parse_uri.return_value = ('s3', 'fake-test-bucket', str(remote_folder))
+
+    _call_convert_text_to_mds(processes=processes,
+                              tokenizer_name=tokenizer_name)
+
+    # Check call counts
+    assert download_and_convert.call_count == processes  # called once per process
+    assert mock_object_store.download_object.call_count == n_text_files + 1  # text files + done file
+    assert mock_object_store.upload_object.call_count == processes + 2  # shard per process + done file + index.json
+
+    if processes > 1:
+        merge_shard_groups.assert_called_once()
+
+    total_object_names = 0
+    for call_args in download_and_convert.call_args_list:
+        object_names = call_args[0][0]
+        total_object_names += len(object_names)
+
+    assert total_object_names == n_text_files  # We should have processed all the text files
+
+    # Check that correct output files exist
+    shards = [f'shard.0000{i}.mds.zstd' for i in range(processes)]
+    _assert_files_exist(prefix=remote_folder,
+                        files=['index.json', get_done_file_name()] + shards)
+
+    _call_convert_text_to_mds(processes=processes,
+                              tokenizer_name=tokenizer_name)
+
+    # Check call counts
+    assert download_and_convert.call_count == processes  # No changes because we shoudn't reprocess
+    assert mock_object_store.download_object.call_count == n_text_files + 2  # One more done file is downloaded
+    assert mock_object_store.upload_object.call_count == processes + 2  # No changes
+
+    # Create an extra text file and call again.
+    n_text_files += 1
+    mock_object_store = Mock(
+        wraps=MockObjectStore(remote_folder, n_text_files, text_content))
+    maybe_create_object_store_from_uri.return_value = mock_object_store
+
+    _call_convert_text_to_mds(processes=processes,
+                              tokenizer_name=tokenizer_name)
+
+    # Check call counts
+    assert download_and_convert.call_count == processes * 2  # called once per process
+    assert mock_object_store.download_object.call_count == n_text_files + 1  # text files + done file
+    assert mock_object_store.upload_object.call_count == processes + 2  # shard per process + done file + index.json
+
+    # Check that the dataset can be used and produces samples with the original text_content
+    tokenizer = AutoTokenizer.from_pretrained(tokenizer_name)
+    dataset = StreamingDataset(local=remote_folder)
```

Comment:
can we iter the whole thing and assert that we have the right amount of data?

Reply:
good idea!